initial firewalld integration tests

maxamillion opened the pull request:
Signed-off-by: Adam Miller <maxamillion@fedoraproject.org>

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Adding a couple test cases to get some baseline integration testing in
for firewalld. If this approach makes sense, I'll add a full set of
test cases to cover the rest of the module's functionality.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
firewalld

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
$ ansible --version          
ansible 2.4.0.0              
  config file = /etc/ansible/ansible.cfg                   
  configured module search path = [u'/home/admiller/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']  
  ansible python module location = /usr/lib/python2.7/site-packages/ansible                                            
  executable location = /usr/bin/ansible                   
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]      
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
$ test/runner/ansible-test integration -v firewalld --allow-destructive
Running firewalld integration test role
Run command: ansible-playbook firewalld-DGDHl_.yml -i inventory -e @integration_config.yml -v
Using /root/ansible/test/integration/integration.cfg as config file
 [WARNING]: The `junit_xml` python module is not installed. Disabling the `junit` callback plugin.


PLAY [testhost] *******************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************
ok: [testhost]

TASK [firewalld : Ensure firewalld is installed] **********************************************************************
ok: [testhost] => {"changed": false, "failed": false, "msg": "", "rc": 0, "results": ["firewalld-0.4.4.4-6.el7.noarch providing firewalld is already installed"]}

TASK [firewalld : firewalld service test permanent enabled] ***********************************************************
changed: [testhost] => {"changed": true, "failed": false, "msg": "Permanent operation, Changed service https to enabled"}

TASK [firewalld : assert firewalld service test permanent enabled worked] *********************************************
ok: [testhost] => {
    "changed": false,
    "failed": false,
    "msg": "All assertions passed"
}

TASK [firewalld : firewalld service test permanent enabled rerun (verify not changed)] ********************************
ok: [testhost] => {"changed": false, "failed": false, "msg": "Permanent operation"}

TASK [firewalld : assert firewalld service test permanent enabled rerun worked (verify not changed)] ******************
ok: [testhost] => {
    "changed": false,
    "failed": false,
    "msg": "All assertions passed"
}

TASK [firewalld : firewalld service test permanent disabled] **********************************************************
changed: [testhost] => {"changed": true, "failed": false, "msg": "Permanent operation, Changed service https to disabled"}

TASK [firewalld : assert firewalld service test permanent disabled worked] ********************************************
ok: [testhost] => {
    "changed": false,
    "failed": false,
    "msg": "All assertions passed"
}

TASK [firewalld : firewalld service test permanent disabled rerun (verify not changed)] *******************************
ok: [testhost] => {"changed": false, "failed": false, "msg": "Permanent operation"}

TASK [firewalld : assert firewalld service test permanent disabled rerun worked (verify not changed)] *****************
ok: [testhost] => {
    "changed": false,
    "failed": false,
    "msg": "All assertions passed"
}

TASK [firewalld : firewalld port test permanent enabled] **************************************************************
changed: [testhost] => {"changed": true, "failed": false, "msg": "Permanent operation, Changed port 8081/tcp to enabled"}

TASK [firewalld : assert firewalld port test permanent enabled worked] ************************************************
ok: [testhost] => {
    "changed": false,
    "failed": false,
    "msg": "All assertions passed"
}

TASK [firewalld : firewalld port test permanent enabled rerun (verify not changed)] ***********************************
ok: [testhost] => {"changed": false, "failed": false, "msg": "Permanent operation"}

TASK [firewalld : assert firewalld port test permanent enabled rerun worked (verify not changed)] *********************
ok: [testhost] => {
    "changed": false,
    "failed": false,
    "msg": "All assertions passed"
}

TASK [firewalld : firewalld port test permanent disabled] *************************************************************
changed: [testhost] => {"changed": true, "failed": false, "msg": "Permanent operation, Changed port 8081/tcp to disabled"}

TASK [firewalld : assert firewalld port test permanent disabled worked] ***********************************************
ok: [testhost] => {
    "changed": false,
    "failed": false,
    "msg": "All assertions passed"
}

TASK [firewalld : firewalld port test permanent disabled rerun (verify not changed)] **********************************
ok: [testhost] => {"changed": false, "failed": false, "msg": "Permanent operation"}

TASK [firewalld : assert firewalld port test permanent disabled rerun worked (verify not changed)] ********************
ok: [testhost] => {
    "changed": false,
    "failed": false,
    "msg": "All assertions passed"
}

PLAY RECAP ************************************************************************************************************
testhost                   : ok=18   changed=4    unreachable=0    failed=0
```

maxamillion committed:
Add a couple test cases to get some baseline integration testing in
for firewalld. If this approach makes sense, I'll add a full set of
test cases to cover the rest of the module's functionality.

mattclay commented on test/integration/targets/firewalld/aliases:
```diff
@@ -0,0 +1,2 @@
+destructive
+posix/ci/group3
```
Add `skip/freebsd` and `skip/osx` here.

mattclay commented:
CI failure in integration tests:

https://app.shippable.com/github/ansible/ansible/runs/41011/36/tests
https://app.shippable.com/github/ansible/ansible/runs/41011/44/tests

maxamillion commented:
@mattclay Thoughts? Does this make sense as an integration test strategy? If so, I'll squash the commits and request this be merged. Then I'll add test cases to cover all features of the module in a later pull request.

maxamillion commented:
@mattclay ping :)

mattclay commented on test/integration/targets/firewalld/tasks/main.yml:
```diff
@@ -0,0 +1,28 @@
+- import_tasks: test_cases.yml
+  when: (ansible_distribution in ["Fedora", "Ubuntu", "Debian", "Arch"] and ansible_python["version"]["major"] == 3) or
+        (ansible_distribution in ["SUSE", "openSUSE Leap"] and ansible_python["version"]["major"] == 2) or
+        (ansible_distribution in ["RedHat", "CentOS"] and ansible_distribution_major_version|int >= 7)
```
Is there a reasonable way to detect the python version used instead of hard-coding the distributions and python versions into the test?

Here is what I see running in CI:

- rhel/7.4
- linux/centos7
- linux/opensuse42.2
- linux/opensuse42.3
- linux/ubuntu1604py3

maxamillion replied:
+1 will do, thanks

maxamillion replied:
@mattclay so I went to fix this up and I remembered the issue isn't what version of python, but what version of python on which distro because some distros only have python2 or python3 versions of firewalld available. Therefore if the tests detect that the `ansible_python_interpreter` is python3, that's not enough information, we unfortunately also have to know what distro/release it's on to know if the test is good to go.

Thoughts? I'm completely open to a better solution, I agree this is ugly.

mattclay replied:
Is it sufficient to perform an `import firewall.config` in a try/except block with the current python interpreter to determine if the test can run?

maxamillion replied:
@mattclay yup, that'll work. :+1:

mattclay reviewed (commented):
Looks good overall. The only concern I have is over the version detection per my other comment.

maxamillion commented:
@mattclay alright, I cleaned up the python2/3 checking, rebased and squashed. Let me know what you think.

mattclay commented:
CI failure in integration tests for:

- centos6: `No package matching 'firewalld' found available, installed or updated`
- ubuntu1404: `firewalld and its python 2 module are required for this module, version 2.0.11 or newer required (3.0.9 or newer for offline operations)`

maxamillion commented:
@mattclay yeah, this is basically why I ended up with the silly conditional I had ... I'm more or less going to end up with a bunch of distro checks anyways. I'll see what makes sense without reintroducing too many distro specific checks.

maxamillion committed:
put the firewalld integration tests in a block for easy distro exclusion

Signed-off-by: Adam Miller <maxamillion@fedoraproject.org>

mattclay commented on test/integration/targets/firewalld/tasks/main.yml:
```diff
@@ -16,15 +16,19 @@
-  when: check_output.rc == 0
+    - import_tasks: test_cases.yml
+      when: check_output.rc == 0
+  when: (ansible_os_family != "RedHat" and ansible_distribution_major_version != "6")
```
Shouldn't this be `when: not (ansible_os_family == "RedHat" and ansible_distribution_major_version == "6")`

maxamillion replied:
@mattclay it 100% should, I didn't see this comment before so mine is slightly different but it's basically the same thing :+1:

maxamillion committed:
refactor, move test cases into their own task import

Signed-off-by: Adam Miller <maxamillion@fedoraproject.org>

maxamillion committed:
fix the conditional to exclude rhel/centos 6

Signed-off-by: Adam Miller <maxamillion@fedoraproject.org>

maxamillion committed:
firewalld - fail offline operations on old lib version

This fixes an issue where the check for an import error would occur
before checking to see if firewalld is in "offline mode" and if it
is, then checking to ensure the version of the firewall python
library was new enough to support offline operations. This patch
will now fail with a correct error message in the scenario that
someone attempts to perform an offline operation but has a version
of the firewall python library that is too old.

Signed-off-by: Adam Miller <maxamillion@fedoraproject.org>

maxamillion commented:
This is now effectively blocked on https://github.com/ansible/ansible/pull/33189 since these tests when run in Shippable found a bug in the firewalld module (running against a distro I'd not tested before).

mattclay commented:
@maxamillion There are unrelated commits in this PR, probably from trying to rebase the branch.

maxamillion commented:
@mattclay yeah, something went wonky with my rebase... will fix.

maxamillion commented:
@mattclay I think I cleaned up after myself. Let me know if there's anything else I should do. I deleted the bot comment that ping'd like 100 people because of the botched rebase.

mattclay commented on test/integration/targets/firewalld/tasks/port_test_cases.yml:
```diff
@@ -0,0 +1,69 @@
+- name: assert firewalld port test permanent enabled worked
+  assert:
+    that:
+    - not result|failed
```
Checking for `failed` is only needed when using `ignore_errors`. There are additional occurrences below.

mattclay commented on test/integration/targets/firewalld/tasks/main.yml:
```diff
@@ -0,0 +1,35 @@
+        state: installed
+      # This doesn't work for CentOS 6 because firewalld doesn't exist in CentOS6
+
+    # Check to make sure the firewalld python module is available.
```
Use `name:` instead of a comment to explain what the task is doing.

maxamillion committed:
don't check for failed, not needed; move comment to name for test explanation

Signed-off-by: Adam Miller <admiller@redhat.com>

mattclay reviewed (approved):
Looks good, aside from the tests as filters CI failures.

maxamillion committed:
don't use filters for tests

Signed-off-by: Adam Miller <admiller@redhat.com>
